mm2src/mm2_main/src/rpc.rs

```diff
@@ -314,62 +318,118 @@ pub extern "C" fn spawn_rpc(ctx_h: u32) {
+                Ok(shutdown_fut) => shutdown_fut,
+                Err(e) => {
+                    error!("MmCtx seems to be stopped already: {e}");
+                    return;
+                },
+            }
+        };
+    }
+
+    // Macro for spawning a server with error handling and logging
+    macro_rules! spawn_server {
+        ($server:expr, $ctx:expr, $ip:expr, $port:expr) => {
+            {
+                let server = $server.then(|r| {
+                    if let Err(err) = r {
+                        error!("{}", err);
+                    };
+                    futures::future::ready(())
+                });
+
+                // As it's said in the [issue](https://github.com/hyperium/tonic/issues/330):
+                //
+                // Aborting the server future will forcefully cancel all connections and not perform a proper drain/shutdown.
+                // While using the special shutdown methods on the server will allow hyper to gracefully drain all connections
+                // and gracefully close connections.
+                common::executor::spawn({
+                    log_tag!(
+                        $ctx,
+                        "😉";
+                        fmt = ">>>>>>>>>> DEX stats {}:{} DEX stats API enabled at unixtime.{} <<<<<<<<<",
+                        $ip,
+                        $port,
+                        gstuff::now_ms() / 1000
+                    );
+                    let _ = $ctx.rpc_started.pin(true);
+                    server
+                });
+            }
+        };
+    }
+
     let ctx = MmArc::from_ffi_handle(ctx_h).expect("No context");
 
-    let rpc_ip_port = ctx.rpc_ip_port().unwrap();
+    let rpc_ip_port = ctx
+        .rpc_ip_port()
+        .unwrap_or_else(|err| panic!("Invalid RPC port: {}", err));
     // By entering the context, we tie `tokio::spawn` to this executor.
     let _runtime_guard = CORE.0.enter();
 
-    let server = Server::try_bind(&rpc_ip_port).unwrap_or_else(|_| panic!("Can't bind on {}", rpc_ip_port));
-    let make_svc = make_service_fn(move |socket: &AddrStream| {
-        let remote_addr = socket.remote_addr();
-        async move {
-            Ok::<_, Infallible>(service_fn(move |req: Request<Body>| async move {
-                let res = rpc_service(req, ctx_h, remote_addr).await;
-                Ok::<_, Infallible>(res)
-            }))
-        }
-    });
-
-    let shutdown_fut = match ctx.graceful_shutdown_registry.register_listener() {
-        Ok(shutdown_fut) => shutdown_fut,
-        Err(e) => {
-            error!("MmCtx seems to be stopped already: {e}");
-            return;
-        },
-    };
-
-    let server = server
-        .http1_half_close(false)
-        .serve(make_svc)
-        .with_graceful_shutdown(shutdown_fut);
+    if ctx.is_https() {
+        // Generate self-signed certificate
+        let subject_alt_names = vec!["localhost".to_string(), "127.0.0.1".to_string()];
```

Comment:
are we allowed to use wildcard(0.0.0.0) address here? 
So I can basically connect to the server from any device connected to my network

Reply:
I think this values should be provided from config file. So people can set their domain and use under tls connections

Reply:
> are we allowed to use wildcard(0.0.0.0) address here?
> So I can basically connect to the server from any device that's connected to my network

mm2 is only meant to be used locally for now, it would be safer to bind it to localhost/127.0.0.1 as the default option and allow other values from config as @ozkanonur suggested.

Using wildcard(0.0.0.0) may inadvertently expose our server to any network interface, which could potentially pose a risk if the machine/server has any vulnerabilities. Therefore, the usage of a wildcard address should be carefully considered based on the purpose of the running mm2 instance, the machine/server security, and the network environment. Using 0.0.0.0 doesn't automatically expose our machine/server to the internet, but it could under certain network configurations (network's firewall and other settings).

Please note that while HTTPS ensures secure and encrypted communications, it doesn't restrict which interfaces can connect to mm2. Hence, using a wildcard address (0.0.0.0) can still expose mm2 to all network interfaces, potentially opening it up to unwanted connections, which HTTPS alone does not prevent.